Search.setIndex({"alltitles": {"1. PhyloTree_CollapseExpand": [[0, "phylotree-collapseexpand"]], "10. GD_Visualizer": [[0, "gd-visualizer"]], "11. GD_Loss_Tracker": [[0, "gd-loss-tracker"]], "12. GD_Loss_Visualizer": [[0, "gd-loss-visualizer"]], "13. Ortho_Retriever": [[0, "ortho-retriever"]], "14. Hybrid_Tracer": [[0, "hybrid-tracer"]], "15. Hybrid_Visualizer": [[0, "hybrid-visualizer"]], "16. HaploFinder": [[0, "haplofinder"]], "2. PhyloSupport_Scaler": [[0, "phylosupport-scaler"]], "3. BranchLength_NumericConverter": [[0, "branchlength-numericconverter"]], "4. Phylo_Rooter": [[0, "phylo-rooter"]], "5. OrthoFilter_LB": [[0, "orthofilter-lb"]], "6. OrthoFilter_Mono": [[0, "orthofilter-mono"]], "7. TreeTopology_Summarizer": [[0, "treetopology-summarizer"]], "8. Tree_Visualizer": [[0, "tree-visualizer"]], "9. GD_Detector": [[0, "gd-detector"]], "Documentation": [[1, "documentation"]], "Features": [[1, "features"]], "Gene To Clade Map": [[2, "gene-to-clade-map"]], "Gene To Expression Map": [[2, "gene-to-expression-map"]], "Gene To Family Map": [[2, "gene-to-family-map"]], "Gene To Order Map": [[2, "gene-to-order-map"]], "Gene To Species Map": [[2, "gene-to-species-map"]], "Gene To Taxa Map": [[2, "gene-to-taxa-map"]], "GeneLength Map": [[2, "genelength-map"]], "Installation": [[3, null]], "Installing PhyloTracer": [[3, "installing-phylotracer"]], "Miniconda": [[3, "id2"]], "Path Map": [[2, "path-map"]], "PhyloTracer Data Files": [[2, null]], "PhyloTracer: A Versatile Toolkit for Comparative Genomics and Phylogenomics Analysis": [[1, null]], "Required Python Packages": [[3, "required-python-packages"]], "Usage": [[0, null]]}, "docnames": ["analyze", "index", "input_files", "installation", "links"], "envversion": {"sphinx": 64, "sphinx.domains.c": 3, "sphinx.domains.changeset": 1, "sphinx.domains.citation": 1, "sphinx.domains.cpp": 9, "sphinx.domains.index": 1, "sphinx.domains.javascript": 3, "sphinx.domains.math": 2, "sphinx.domains.python": 4, "sphinx.domains.rst": 2, "sphinx.domains.std": 2}, "filenames": ["analyze.rst", "index.rst", "input_files.rst", "installation.rst", "links.rst"], "indexentries": {}, "objects": {}, "objnames": {}, "objtypes": {}, "terms": {"0": [0, 2, 3], "0001": 0, "0005": 0, "07g273800": 2, "1": [2, 3], "100": 0, "12": 2, "1467": 2, "2": [2, 3], "201": 2, "3": 3, "3417": 2, "5": 2, "50": 0, "6": 3, "A": [0, 3], "By": 1, "For": 0, "If": [0, 2], "It": 0, "The": [0, 2, 3], "There": 3, "These": 3, "To": [0, 3], "With": 3, "abab": [0, 1], "abba": 1, "absolut": 0, "absolute_branch_length": 0, "accomplish": 3, "accord": 2, "accur": [0, 1], "accuraci": 1, "across": 0, "activ": 3, "add": 2, "addit": 0, "addition": 1, "adjust": 0, "affect": 0, "after": 0, "align": 0, "all": [0, 3], "allow": 0, "also": 0, "altern": 3, "amborella_trichopoda": 2, "amborellacea": 2, "amborellal": 2, "amtr_s00796p00010580": 2, "an": 0, "analysi": 0, "analyt": 0, "analyz": 0, "angiosperm": 2, "ani": [0, 2], "annot": 2, "answer": 3, "applic": 3, "ar": [2, 3], "arabidopsis_thaliana": 2, "ask": 3, "atcg00500": 2, "avail": 3, "b": 0, "baba": 0, "base": 0, "bash": 3, "becaus": 3, "befor": [0, 3], "below": [2, 3], "between": [0, 1], "binari": 0, "bootstrap": 0, "both": 0, "branch": [0, 1], "branch_length_multipl": 0, "brassicacea": 2, "brassical": 2, "burst": 0, "c": 0, "can": [0, 2, 3], "care": 0, "caution": 0, "cd": 3, "chmod": 3, "chr_a": 0, "chr_b": 0, "chromosom": 0, "circular": 0, "clade": [0, 1], "classif": 0, "click": 3, "clone": [2, 3], "collaps": 0, "com": 3, "comb": 0, "command": [0, 3], "compat": 0, "comprehens": 1, "comput": 0, "concept": 1, "conda": 3, "constraint": 0, "contain": 0, "context": 0, "continuum": 3, "conveni": 3, "convers": 0, "convert": 0, "copi": [0, 1], "correspond": 0, "could": 3, "coverag": 0, "creat": 3, "curl": 3, "d": 0, "data": 1, "decim": 0, "decimal_plac": 0, "deepvar": 0, "default": 0, "depth": [0, 1], "describ": 2, "descript": 0, "design": 1, "detail": [0, 2], "detect": [0, 1], "determin": 1, "directori": 2, "distanc": 1, "distinguish": 0, "distribut": 3, "dna": 2, "do": [2, 3], "doc": 3, "document": 0, "dotplot": 0, "download": 3, "downstream": 0, "dup_species_num": 0, "dup_species_proport": 0, "duplic": [0, 1], "dynam": 0, "e": 0, "each": [0, 2], "easier": 3, "egl": 3, "eglf": 3, "elimin": 1, "en": 3, "enabl": [0, 1], "end": 0, "end_speci": 0, "enhanc": [0, 1], "ensur": 0, "enumer": 0, "env": 3, "environ": 3, "equal": 0, "establish": 0, "ete3": 3, "even": 3, "event": 0, "everi": 3, "evolutionari": 0, "exampl": [0, 2], "example_data": 2, "exceed": 0, "execut": 3, "export": 3, "express": 0, "f": 3, "fabacea": 2, "fabal": 2, "fabid": 2, "facilit": 1, "fail": 3, "famili": 0, "featur": 0, "file": [0, 1], "filter": [0, 1], "filtrat": 1, "find": 0, "fix": [2, 3], "folder": 2, "follow": 3, "form": 0, "format": [0, 1], "fould": 1, "found": [2, 3], "foundat": 1, "frequenc": 0, "friendli": 1, "from": [0, 2, 3], "fulli": 0, "g": 0, "gd": [0, 1], "gd_result": 0, "gd_support": 0, "gene": [0, 1], "gene2clad": [0, 2], "gene2express": [0, 2], "gene2famili": [0, 2], "gene2length": 0, "gene2ord": [0, 2], "gene2sp": [0, 2], "gene2taxa": 2, "gene_categori": 0, "gene_express": 0, "gene_famili": 0, "genom": 0, "get": 3, "gf": 0, "gf_id2path": [0, 2], "gff": 0, "git": 3, "github": [2, 3], "glx": 3, "glycine_max": 2, "glyma": 2, "graph": 0, "greater": 0, "group": 0, "guid": 3, "haplotyp": 0, "have": [2, 3], "heatmap": 0, "here": 2, "highlight": 0, "how": [0, 3], "html": 3, "http": 3, "hybrid": [0, 1], "hyde": [0, 3], "hyde_out": 0, "i": [0, 1, 3], "identif": 1, "identifi": 0, "imap": [0, 2], "includ": [0, 1, 3], "incorpor": 1, "index": [0, 3], "individu": 2, "infer": [0, 1], "inform": 0, "ingroup": 1, "initi": 3, "input": 2, "input_gene_length": 0, "input_gf_list": 0, "input_imap": 0, "input_seq_gf_list": 0, "input_sps_tre": 0, "input_taxa": 0, "insert": [0, 1], "insert_branch_index": 0, "instal": [1, 2], "install_packag": 3, "instruct": 0, "integr": 0, "interest": 0, "introduc": 1, "io": 3, "issu": 3, "keep_branch": 0, "lai": 1, "larg": 0, "latest": 3, "len": 0, "length": [0, 1], "less": 0, "level": 0, "light": 0, "limit": 0, "line": 0, "linux": 3, "linuxfb": 3, "list": [0, 3], "ll": 3, "load": 3, "long": [0, 1], "look": 3, "loss": [0, 1], "mac": 3, "macosx": 3, "mai": 3, "main": 2, "make": 3, "malvid": 2, "manag": 3, "manipul": 1, "manual": 3, "map": 0, "mark": 0, "matplotlib": 3, "maxim": 1, "maximum": 0, "meet": 0, "member": 0, "method": 0, "miniconda3": 3, "minim": [1, 3], "minimalegl": 3, "model": [0, 1], "modul": [0, 3], "monophylet": [0, 1], "more": [1, 2], "multi": 0, "multipl": 0, "multipli": 0, "n1": 0, "name": 0, "need": [0, 3], "newick": 0, "next": 3, "nitrogen": 2, "node": 0, "nois": [0, 1], "non": 2, "number": [0, 2], "numbered_species_tre": 0, "numer": 0, "numpi": 3, "nwk": 0, "o": 3, "object": 0, "offscreen": 3, "og_1": 2, "og_2": 2, "og_3": 2, "onc": 3, "one": [0, 3], "onli": 0, "onto": 0, "oper": 3, "option": 0, "order": 0, "origin": 1, "ortholog": [0, 1], "osx": 3, "otherwis": 0, "out": 0, "outgroup": 1, "outlier": 0, "overlap": [0, 1], "panda": 3, "paralog": 0, "paramet": 0, "path": 0, "per": 0, "phylo_root": 2, "phylo_trac": 0, "phylogenet": [0, 1], "phylogenom": 0, "phylotrac": 0, "pip": 3, "place": 0, "platform": 3, "pleas": 3, "plugin": 3, "point": 0, "popul": 2, "posterior": 0, "potenti": [0, 3], "pphylotrac": 3, "predefin": 0, "preserv": 0, "principl": 1, "priorit": 0, "probabl": 0, "problem": 3, "profil": 0, "project": 3, "proport": 0, "provid": [0, 1, 2], "prune": 0, "put": [0, 1], "py": 0, "pypdf4": 3, "pypi": [2, 3], "pyqt5": 3, "qpa": 3, "qt": 3, "qt_qpa_platform": 3, "question": 3, "r": 0, "rang": 0, "ratio": [0, 1], "recalibr": 0, "recommend": 3, "reconcil": 0, "rectangular": 0, "reduc": 1, "refer": 0, "reinstal": 3, "rel": 0, "relationship": 0, "relative_branch_length": 0, "relev": 0, "remind": 3, "remov": 0, "repo": 3, "repositori": 2, "requir": 0, "resolv": 0, "respect": [0, 2], "result": 0, "retriev": 1, "revert": 0, "rf": 1, "right": 2, "robinson": 1, "root": [0, 1], "run": [2, 3], "sampl": 2, "scale": 0, "scale_to": 0, "score": 1, "section": 0, "select": 0, "seq_gf_id2path": 0, "sequenc": [0, 2], "sequenti": 2, "sh": 3, "should": 0, "signal": [0, 1], "singl": [0, 1], "size": 0, "solid": 1, "speci": [0, 1], "species_a": 0, "species_a_gff": 0, "species_a_len": 0, "species_b": 0, "species_b_gff": 0, "species_b_len": 0, "specif": 0, "specifi": 0, "split": 0, "sptree": 0, "squar": 0, "stabl": 3, "stack": 0, "start": 3, "start_nod": 0, "statist": [0, 1], "step": 3, "string": 0, "structur": 0, "style": 0, "subclad": 0, "subclade_support": 0, "subgenom": 0, "subtre": 0, "summar": [0, 1], "summari": 0, "superimposit": 0, "support": 0, "support_valu": 0, "system": 3, "tag": 0, "take": 3, "target_nod": 0, "taxa": 0, "taxonom": 0, "test": 0, "than": 0, "thei": 2, "them": 0, "thi": [0, 3], "though": 3, "threshold": 0, "through": [0, 3], "time": 3, "tip": [0, 2], "tool": 1, "topologi": [0, 1], "tqdm": 3, "trace": 0, "transform": 0, "tree": [0, 1], "tree_styl": 0, "treefil": 2, "two": 3, "txt": 0, "type": [1, 3], "under": 0, "unless": 0, "up": 0, "us": [0, 1, 3], "usag": 1, "user": [1, 3], "v3": 3, "valu": 0, "varianc": [0, 1], "variou": 0, "verifi": 1, "versa": 0, "vice": 0, "view": 2, "visual": [0, 1, 3], "visual_chr_a": 0, "visual_chr_b": 0, "visual_gd": 0, "visualiz": 0, "vnc": 3, "wa": 3, "wai": 3, "wayland": 3, "we": [2, 3], "webgl": 3, "well": 0, "were": 0, "where": 0, "whether": 0, "whose": 0, "within": 2, "x": 3, "x86_64": 3, "xcb": 3, "xcomposit": 3, "ye": 3, "yiyongzhao": 3, "yml": 3, "you": [2, 3], "your": 3}, "titles": ["Usage", "PhyloTracer: A Versatile Toolkit for Comparative Genomics and Phylogenomics Analysis", "PhyloTracer Data Files", "Installation", "&lt;no title&gt;"], "titleterms": {"1": 0, "10": 0, "11": 0, "12": 0, "13": 0, "14": 0, "15": 0, "16": 0, "2": 0, "3": 0, "4": 0, "5": 0, "6": 0, "7": 0, "8": 0, "9": 0, "A": 1, "To": 2, "analysi": 1, "branchlength_numericconvert": 0, "clade": 2, "compar": 1, "data": 2, "document": 1, "express": 2, "famili": 2, "featur": 1, "file": 2, "gd_detector": 0, "gd_loss_track": 0, "gd_loss_visu": 0, "gd_visual": 0, "gene": 2, "genelength": 2, "genom": 1, "haplofind": 0, "hybrid_trac": 0, "hybrid_visu": 0, "instal": 3, "map": 2, "miniconda": 3, "order": 2, "ortho_retriev": 0, "orthofilter_lb": 0, "orthofilter_mono": 0, "packag": 3, "path": 2, "phylo_root": 0, "phylogenom": 1, "phylosupport_scal": 0, "phylotrac": [1, 2, 3], "phylotree_collapseexpand": 0, "python": 3, "requir": 3, "speci": 2, "taxa": 2, "toolkit": 1, "tree_visu": 0, "treetopology_summar": 0, "usag": 0, "versatil": 1}})